File: app/src/test/java/net/reichholf/dreamdroid/enigma/ServiceParserTest.java
package net.reichholf.dreamdroid.enigma;

import org.junit.Test;

import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

public class ServiceParserTest {
    @Test
    public void parsesGetServicesFixtureIntoServiceValues() throws Exception {
        InputStream in = getClass().getResourceAsStream("/web/getservices.xml");
        assertNotNull(in);
        String xml = new String(in.readAllBytes(), StandardCharsets.UTF_8);
        long started = System.nanoTime();
        List<Service> services = ServiceParser.INSTANCE.parse(xml);
        long elapsedNanos = System.nanoTime() - started;
        System.out.println("ServiceParser.parse nanos=" + elapsedNanos);

        assertEquals(3, services.size());
        assertEquals(
                "1:7:1:0:0:0:0:0:0:0:FROM BOUQUET \"userbouquet.favourites.tv\" ORDER BY bouquet",
                services.get(0).getReference()
        );
        assertEquals("Favourites (TV)", services.get(0).getName());
        assertEquals("1:0:1:6DCA:44D:1:C00000:0:0:0:", services.get(1).getReference());
        assertEquals("Das Erste HD", services.get(1).getName());
        assertEquals("1:64:1:0:0:0:0:0:0:0:", services.get(2).getReference());
        assertEquals("--------", services.get(2).getName());
    }
}

File: app/test/java/net/reichholf/dreamdroid/enigma/ServiceParserTest.kt
package net.reichholf.dreamdroid.enigma

import net.reichholf.dreamdroid.testutil.loadWebFixture
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class ServiceParserTest {
    @Test
    fun parsesGetServicesFixtureIntoServiceValues() {
        val xml = loadWebFixture("getservices.xml")
        val started = System.nanoTime()
        val services = ServiceParser.parse(xml)
        println("ServiceParser.parse nanos=${System.nanoTime() - started}")

        assertEquals(3, services.size)
        assertEquals(
            "1:7:1:0:0:0:0:0:0:0:FROM BOUQUET \"userbouquet.favourites.tv\" ORDER BY bouquet",
            services[0].reference
        )
        assertEquals("Favourites (TV)", services[0].name)
        assertEquals("1:0:1:6DCA:44D:1:C00000:0:0:0:", services[1].reference)
        assertEquals("Das Erste HD", services[1].name)
        assertEquals("1:64:1:0:0:0:0:0:0:0:", services[2].reference)
        assertEquals("--------", services[2].name)
    }

    @Test
    fun emptyXmlYieldsNoServices() {
        assertEquals(0, ServiceParser.parse("").size)
    }

    @Test
    fun malformedXmlYieldsNoServices() {
        assertEquals(0, ServiceParser.parse("<e2servicelist><e2service>").size)
    }

    @Test
    fun stripsControlCharactersFromServiceName() {
        val xml = """
            <?xml version="1.0" encoding="UTF-8"?>
            <e2servicelist>
            <e2service>
            <e2servicereference>1:0:1:1:1:1:0:0:0:0:</e2servicereference>
            <e2servicename>ZDF\u0001HD</e2servicename>
            </e2service>
            </e2servicelist>
        """.trimIndent().replace("\\u0001", "\u0001")
        val services = ServiceParser.parse(xml)
        assertEquals(1, services.size)
        assertEquals("ZDFHD", services[0].name)
    }
}
